File: dotnet/Statistics/Statistics/Frequency.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Statistics
{
    /// <summary>
    /// Represents the frequency statistics.
    /// </summary>
    internal class Frequency<TValue> where TValue : class
    {
        private readonly IDictionary<IComparable<TValue>, long> _frequencyValues;

        private readonly IList<IComparable<TValue>> _modeValues;

        private long _maxFrequency;

        internal Frequency()
        {
            _frequencyValues = new Dictionary<IComparable<TValue>, long>();
            _modeValues = new List<IComparable<TValue>>();
        }

        internal void AddValue(IComparable<TValue> value)
        {
            if (_frequencyValues.ContainsKey(value))
            {
                var frequency = ++_frequencyValues[value];
                if (_maxFrequency == frequency)
                {
                    _modeValues.Add(value);
                }
                else if (_maxFrequency < frequency)
                {
                    _maxFrequency = frequency;
                    _modeValues.Clear();
                    _modeValues.Add(value);
                }
            }
            else
            {
                _frequencyValues.Add(value, 1);
                if (_maxFrequency == 1)
                {
                    _modeValues.Add(value);
                }
                else if (_maxFrequency < 1)
                {
                    _maxFrequency = 1;
                    _modeValues.Add(value);
                }
            }
        }

        internal long GetFrequencyCount(IComparable<TValue> value)
        {
            if (_frequencyValues.ContainsKey(value))
            {
                return _frequencyValues[value];
            }

            return 0;
        }

        internal IList<IComparable<TValue>> ModeValues
        {
            get
            {
                return _modeValues;
            }
        }
    }
}

File: cplusplus/Statistics/Statistics/Frequency.cpp
#include "stdafx.h"
#include "Frequency.h"

namespace edu
{
	namespace statistics
	{

		Frequency::Frequency() : _maxFrequency(0)
		{
		}


		Frequency::~Frequency()
		{
		}


		void Frequency::addValue(wstring &value)
		{
			map<wstring, size_t>::iterator iterator = _frequencyValues.find(value);
			if (_frequencyValues.end() != iterator)
			{
				size_t frequency = ++(iterator->second);
				if (_maxFrequency == frequency)
				{
					_modeValues.push_back(value);
				}
				else if (_maxFrequency < frequency)
				{
					_maxFrequency = frequency;
					_modeValues.clear();
					_modeValues.push_back(value);
				}
			}
			else
			{
				_frequencyValues[value] = 1;
				if (_maxFrequency == 1)
				{
					_modeValues.push_back(value);
				}
				else if (_maxFrequency < 1)
				{
					_maxFrequency = 1;
					_modeValues.push_back(value);
				}
			}
		}


		size_t Frequency::frequencyCount(wstring &value)
		{
			map<wstring, size_t>::iterator iterator = _frequencyValues.find(value);
			if (_frequencyValues.end() != iterator)
			{
				return iterator->second;
			}

			return 0;
		}


		vector<wstring>& Frequency::modeValues()
		{
			return _modeValues;
		}
	}
}